### PR TITLE
ValidateMoabJob added 

### LINKED ISSUE #1697: make it possible to checksum validate a Moab without having to reference preservation catalog's database
⚠️ _warning_ ⚠️ -- it doesn't matter whether this happens before or after #1696, just make sure both aren't being worked at the same time, because then one or more developers will likely be in for a painful rebase.

see #1691 for more context.

Hopefully it isn't _too_ difficult to do this by slightly refactoring and generalizing `ChecksumValidator` so that it can take a druid and a storage location, or even just a path to a Moab, and do all the same checksum validation it currently does, without having to look at or update the database.  It should make results available to the caller.  While it would be convenient to point `ChecksumValidator` at any purported Moab folder in any location, for practical purposes for now, it is safe to assume that the Moab in question at least resides in the known storage roots in a Moab directory tree (since we are still validating content being updated by the usual pres robots process).

I'm happy to pair on implementation ideas if that's helpful, and also jotted down some thoughts around implementation here:  https://github.com/sul-dlss/preservation_catalog/issues/1691#issuecomment-841437740

Maybe it _should_ mark the `CompleteMoab#status` as `invalid_checksum` if there is an entry in the catalog already, but it _shouldn't_ create an entry anew, or update anything else (including `#version`)?  Not sure, warrants further discussion.  Would be fine to err on the side of being naive and not updating at all.  Another option would be to set the `PreservedObject#robot_versioning_allowed` flag to `false` if there is already a `PreservedObject` record.

### PR DESCRIPTION
**Pull request by ndushay**
## Why was this change made?

In order to avoid continuing with preservationIngestWF when a Moab hasn't been correctly written to Ceph, or isn't yet ready to be accessed by the replication steps, we want to add a validation of the on disk Moab, without reference to the preservation-catalog basis.

Closes #1697, #1698

Review the 2 commits separately;  the first is a rename for clarity;  the second accomplishes the goals of the tickets.

## How was this change tested?

- unit tests
- script that is similar

### CODE REVIEW

File: spec/jobs/validate_moab_job_spec.rb [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        allow(storage_object_version1).to receive(:verify_version_additions).and_call_original
+        allow(storage_object_version1).to receive(:verify_version_inventory).and_call_original
+        allow(storage_object_version2).to receive(:verify_signature_catalog).and_call_original
+        allow(storage_object_version2).to receive(:verify_manifest_inventory).and_call_original
+        allow(storage_object_version2).to receive(:verify_version_additions).and_call_original
+        allow(storage_object_version2).to receive(:verify_version_inventory).and_call_original
+        allow(storage_object_version3).to receive(:verify_signature_catalog).and_call_original
+        allow(storage_object_version3).to receive(:verify_manifest_inventory).and_call_original
+        allow(storage_object_version3).to receive(:verify_version_additions).and_call_original
+        allow(storage_object_version3).to receive(:verify_version_inventory).and_call_original
```
Comment (jmartin-sul) [inline spec/jobs/validate_moab_job_spec.rb:81]: i'm _slightly_ uncomfortable with how closely tied this section of tests is to the current implementation (in terms of the new class' calls to moab-versioning methods and in terms of assumptions about calls made by those moab-versioning methods).

on the other hand, we know those methods do what we want, they're not super likely to change soon, and the app code does call them out in the comments.

i think i was hoping that the tests would make more use of the existing fixture objects with known problems, but looking back at `checksum_validator_spec.rb`, it does more of this sort of mocking and reaching into moab-versioning than i remembered, and it's a little less clear about which druids have which issues than i'd hoped/remembered.

still... how would you feel about adding a test or two that just runs on a problematic moab, mocking none of the moab-versioning calls, with some expectations about what specific error messages get sent to WF service?  maybe even just a more exacting message expectation in the test on line 47?  and maybe one good moab resulting in marking the WF step complete?

for example, i think `zz111rr1111` has a checksum mismatch for `v0001/data/content/eric-smith-dissertation-augmented.pdf`, `tt222tt2222` is missing a content file, `zz333vv3333` is missing a sigCat, and `bz514sm9647` is ok?  that's my read from looking at the other checksum validator spec for a bit.

regardless, i don't think i'd block merging over this point.  i believe this code works and pushes the ball forward, just worried about such tightly coupled tests making it harder to refactor later.

**Review by jmartin-sul — CHANGES_REQUESTED** [review]
one comment that i think needs fixing for accuracy, one question about a difference between two `rescue`s, an error handling suggestion, and a testing worry.

i think fixing the comment (or explaining to me that i'm wrong about the comment 🙂) is the only thing over which i'd actually block merging.  the other considerations feel to me like they'd improve things, but i don't think they address anything i'd call "wrong."

File: app/jobs/validate_moab_job.rb [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    errors << structural_errors unless structural_errors.empty?
+
+    moab.version_list.each do |version|
+      # ensure all files in signatureCatalog.xml exist
+      errors << verification_errors(version.verify_signature_catalog)
+
+      # verify_version_storage includes:
+      #   verify_manifest_inventory, (which computes and compares v000x/manifest file checksums)
+      #   verify_version_inventory,
+      #   verify_version_additions (which computes and compares v000x/data/metadata file checksums)
```
[the remark this answers is not in the document]
Comment (ndushay) [reply app/jobs/validate_moab_job.rb:46]: thanks!  missed this.  It's probably also commented incorrectly in moab-versioning README and wiki - I'll take a look.

File: app/jobs/validate_moab_job.rb [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    start = Time.zone.now
+    log_started
+
+    # TODO: what if we lose connectivity while it's running in resque
+    errors = validate
+    if errors.empty?
+      log_success(Time.zone.now - start)
+    else
+      log_failure(errors)
+    end
```
[the remark this answers is not in the document]
Comment (ndushay) [reply app/jobs/validate_moab_job.rb:26]: happy to do this.  was following the pattern in https://github.com/sul-dlss/technical-metadata-service/blob/main/app/jobs/technical_metadata_workflow_job.rb without thinking too much.

File: app/jobs/validate_moab_job.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+      # verify_version_storage includes:
+      #   verify_manifest_inventory, (which computes and compares v000x/manifest file checksums)
+      #   verify_version_inventory,
+      #   verify_version_additions (which computes and compares v000x/data/metadata file checksums)
+      errors << verification_errors(version.verify_version_storage)
+    rescue Errno::ENOENT => e
+      errors << e.message # No such file or directory
+    rescue Nokogiri::XML::SyntaxError => e
+      errors << e
```
[the remark this answers is not in the document]
Comment (ndushay) [reply app/jobs/validate_moab_job.rb:51]: Errno:ENOENT is unintelligible to most of us, and the message text is clear "no such file or directory".   In the latter case, the error message is sometimes something like "the file is empty" and having the name of the error included is helpful information.

File: spec/jobs/validate_moab_job_spec.rb [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+      expect(workflow_client).to have_received(:update_status).with(druid: druid,
+                                                                    workflow: 'preservationIngestWF',
+                                                                    process: 'validate-moab',
+                                                                    elapsed: a_value > 0,
+                                                                    status: 'completed',
+                                                                    note: a_string_starting_with(exp_str))
+    end
+
+    it 'reports failure to workflow server when there are validation errors' do
+      job.perform(druid)
```
[the remark this answers is not in the document]
Comment (ndushay) [reply spec/jobs/validate_moab_job_spec.rb:48]: correct.  It turns out the fixture object doesn't pass validation.  I used what we had.  :-)

File: spec/jobs/validate_moab_job_spec.rb (hunk 1 above)
Comment (ndushay) [reply spec/jobs/validate_moab_job_spec.rb:81]: In a perfect world, I agree with you.  I took the expeditious route.  As you note, these tests aren't about showing that moab-validation can trap errors, they're about showing that we actually perform checksum validations, using the moab-val methods.

I did think about your point when I was writing the tests, and felt that since I was testing that the moab-val methods which do the actual comparison work were called (and not the moab-val method that is called in the job) ... that this was a reasonable middle ground.

Checking for specific errors ... could end up depending on the strings used by moab-val for error messages, or on specific errors thrown.   At some point, you have to trust that the tools you're using ... work.   If moab-val changes significantly, I think we'll know that prescat will need work.

File: app/jobs/validate_moab_job.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmartin-sul) [reply app/jobs/validate_moab_job.rb:26]: thanks!

File: app/jobs/validate_moab_job.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmartin-sul) [reply app/jobs/validate_moab_job.rb:53]: makes sense, thanks for the explanation, sorry to second guess!

**Comment by ndushay** [comment]
to heck with code climate;  the only thing not covered in new code is raising an argument error when no druid and rescuing StandardError.

File: spec/jobs/validate_moab_job_spec.rb (hunk 1 above)
Comment (jmartin-sul) [reply spec/jobs/validate_moab_job_spec.rb]: indeed

**Review by jmartin-sul — APPROVED** [review]
thanks for the changes!  approved, but leaving un-merged since you said you wanted to squash the last two commits together.

**Comment by jmartin-sul** [comment]
> to heck with code climate; the only thing not covered in new code is raising an argument error when no druid and rescuing StandardError.

👍  yeah, not worried about those
